---
title: "Internal Citrobacter Tree w-refs"
author: "Max Cummins"
date: "17/05/2022"
output:
    html_document:
        theme: spacelab
        toc: true
        toc_float: true
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)

#Read in our packages
library(tidyverse)
library(abricateR)
library(dplyr)
library(ggplot2)
library(ggtree)
library(pheatmap)
library(ComplexHeatmap)
library(phytools)
#library(plotly)
library(readr)
library(readxl)
#library(paletteer)
library(grid)
library(gridExtra)
library(hrbrthemes)
library(cowplot)
#library(plasmidmapR)

#Define our not in function
`%nin%` <- Negate(`%in%`)

```

## Outline


## File summary

A GitHub repository available at *https://github.com/maxlcummins/*
contains all scripts and files required to rerun analysis detailed in this
document.

Below are the files associated with this research project and report.

```
File tree
```

```{r echo=FALSE}
#Load paths to files needed for the script
abricate_path <- "results/summaries/genotype.txt"
pMLST_path <- "results/summaries/pMLST.txt"
mlst_path <- "results/summaries/mlst.txt"
tree_path <- "results/summaries/trees/internal_tree_w_refseq_refs.tree"

species_ids_path <- "bacsort/Species_ids.csv"

#Provide output names
output_name <- "Citrobacter"

#Provide output directory name
output_dir <- "output"

#Create an output directory
dir.create(file.path(getwd(), "output"), showWarnings = FALSE)
```



```{r read_in_public_metadata, echo=FALSE, include=FALSE}
metadata <- read_excel("delims/Metadata_public_genomes.xlsx", 
    sheet = "metadata_only")

```


```{r quiet_abricate, echo=FALSE,include=FALSE}
#Run AbricateR to summarise genomic data
abricateR::abricateR(
        abricate_in = abricate_path,
        output = output_name,
        output_directory = output_dir,
        identity = 90,
        length = 90,
        writecsv = FALSE,
        #pMLST_data = pMLST_path
)

#Read in our species IDs
species_ids <- read_csv("bacsort/Species_ids.csv")

#Create a new column for our collection_names
species_ids <- species_ids %>%
        mutate(collection = case_when(grepl("GCF", name) ~ "RefSeq",
                                      grepl("SD", name) ~ "Gull",
                                      grepl("SG", name) ~ "Gull",
                                      grepl("ERQ", name) ~ "Wastewater",
                                      grepl("DB", name) ~ "Wastewater"))
                                      

#Create our metadata sheet (we will expand on this later with extra data)
metadata <- left_join(species_ids, metadata)

#Fix the source column for internal isolates
metadata <- metadata %>% mutate(Source = if_else(collection == "RefSeq", Source, collection))

#Read in the list of genomes which passed QC
QC_pass <- read_csv("delims/QC_pass.txt", col_names = "name", show_col_types = FALSE)

```


```{r MLST_placeholders, echo=FALSE}
#Read in MLST data
mlst <- read_delim(
    mlst_path,
    delim = "\t",
    escape_double = FALSE,
    trim_ws = TRUE,
    show_col_types = FALSE
)

#Filter out samples which didnt pass QC
mlst <- mlst %>% filter(name %in% QC_pass$name)

#Produce placeholder STs
placeholder_STs <- mlst %>% select(-name, -scheme) %>% unique() %>% filter(ST == "-") %>% select(-ST) %>%
       # filter(
       #         !grepl("-", allele1),
       #         !grepl("-", allele2),
       #         !grepl("-", allele3),
       #         !grepl("-", allele4),
       #         !grepl("-", allele5),
       #         !grepl("-", allele6),
       #         !grepl("-", allele7)) %>% 
mutate(ST_new = paste0("placeholder_ST",as.character(1:nrow(.)))) %>% mutate(allele_cat = paste0(allele1, allele2, allele3, allele4, allele5, allele6, allele7)) %>% select(ST_new, allele_cat)

#Produce a list of the mlsts within the collection
ST_list <- mlst %>% mutate(allele_cat = paste0(allele1, allele2, allele3, allele4, allele5, allele6, allele7)) %>% left_join(placeholder_STs, by = "allele_cat") %>% select(-allele_cat) %>% mutate(ST_new = if_else(is.na(ST_new), ST, ST_new))


#Set our seed for reproducibility
set.seed(2)

#Generate a random sample (n=1) of strains grouped by species
sample_for_tree <- species_ids %>% inner_join(ST_list, by = "name") %>% filter(name %in% QC_pass$name) %>% filter(grepl("GCF",name)) %>% group_by(bacsort_species) %>% slice_sample(n=1)

#Generate a second random sample (n=1) of strains grouped by species
#Note we remove strains with the same STs as we picked previously
sample_for_tree_2 <- species_ids  %>% inner_join(ST_list, by = "name") %>% filter(name %in% QC_pass$name) %>% filter(ST_new %nin% sample_for_tree$ST_new) %>% filter(grepl("GCF",name)) %>% group_by(bacsort_species) %>% slice_sample(n=1)

#Bind the list of the genomes selected by sample one and two
sample_for_tree <- bind_rows(sample_for_tree, sample_for_tree_2)

#Same again for another 3 iterations...
sample_for_tree_3 <- species_ids %>% inner_join(ST_list, by = "name") %>% filter(name %in% QC_pass$name) %>% filter(ST_new %nin% sample_for_tree$ST_new) %>% filter(grepl("GCF",name)) %>% group_by(bacsort_species) %>% slice_sample(n=1)

sample_for_tree <- bind_rows(sample_for_tree, sample_for_tree_3)

sample_for_tree_4 <- species_ids %>% inner_join(ST_list, by = "name") %>% filter(name %in% QC_pass$name) %>% filter(ST_new %nin% sample_for_tree$ST_new) %>% filter(grepl("GCF",name)) %>% group_by(bacsort_species) %>% slice_sample(n=1)

sample_for_tree <- bind_rows(sample_for_tree, sample_for_tree_4)

sample_for_tree_5 <- species_ids %>% inner_join(ST_list, by = "name") %>% filter(name %in% QC_pass$name) %>% filter(ST_new %nin% sample_for_tree$ST_new) %>% filter(grepl("GCF",name)) %>% group_by(bacsort_species) %>% slice_sample(n=1)

sample_for_tree <- bind_rows(sample_for_tree, sample_for_tree_5)

#Write our list of samples to file
#write_delim(sample_for_tree, "delims/sample_of_reference_genomes.txt", delim = "\t")

#Bind the metadata and the new ST placeholder names
metadata <- metadata %>% right_join(ST_list, by = "name")

```

```{r, novel_MLSTs, include = FALSE, echo = FALSE}
#Read in novel MLST data
pubmlst <- read_delim(
    "delims/pubMLST.txt",
    delim = "\t",
    escape_double = FALSE,
    trim_ws = TRUE,
    show_col_types = FALSE
)

#Filter out samples which didn't pass QC
pubmlst <- pubmlst %>% rename('name' = isolate) %>% filter(name %in% QC_pass$name) %>% select(name, arcA:mdh,`ST (MLST)`, -matches("BACT"))

#Change column names
colnames(pubmlst) <- c('name', 'allele5', 'allele1', 'allele2', 'allele6', 'allele3', 'allele7', 'allele4', 'ST')

pubmlst <- pubmlst %>% mutate(across(everything(), ~ as.character(.))) %>% rename('ST_pub' = ST) %>% mutate(ST_pub = str_replace(ST_pub, 'NA', 'Untypeable')) %>% select(name, ST_pub)

placeholder_to_new <- left_join(ST_list, pubmlst) %>% as.data.frame() %>% select(ST_new, ST_pub) %>% unique()

ST_list <- left_join(ST_list, placeholder_to_new, by = 'ST_new', relationship = 'many-to-many')

ST_list <- ST_list %>% group_by(name) %>% slice_max(order_by = ST_pub, n = 1)

ST_list <- ST_list %>% mutate(ST_placeholder = ST_new)

ST_list <- ST_list %>% mutate(ST_new = if_else(is.na(ST_pub), ST_new, ST_pub))
```


```{r make_tree, include=FALSE, echo=FALSE}

#Read in our tree
tree <- read.tree(tree_path)

#Midpoint root our tree
tree <- midpoint.root(tree)

tree$tip.label <- gsub("Citrobacter_", "C.", tree$tip.label)

df <- as.data.frame(tree$tip.label, tree$tip.label) %>% rownames_to_column('name') %>% rename('bacsort_species' = `tree$tip.label`)

df <- df %>% mutate(bacsort_species = str_replace(pattern = '_.*', replacement = '', bacsort_species)) %>%
        mutate(bacsort_species = str_replace(pattern = 'C.', replacement = 'Citrobacter ', bacsort_species)) %>%
        filter(!grepl("^ERQ|^S|^DB", name))

missings <- df %>% filter(bacsort_species %nin% metadata$bacsort_species) %>% pull(bacsort_species)

metadata <- bind_rows(metadata, df)

#Define our colours list for our species names
#species_cols <- c('blue', 'cyan', 'green', '#AAAA00', 'red', 'purple', 'black') #AAAA00 = olive


#species_cols <- c(#Red
#"#e6194B",
##Green
#"#3cb44b",
###Yellow
##"#ffe119",
##Blue
#"#4363d8",
##Orange
#"#f58231",
##Purple
#"#911eb4",
##Cyan
#"#42d4f4",
##Magenta
#"#f032e6",
##Lime
#"#bfef45",
##Pink
#"#fabed4",
##Teal
#"#469990",
###Lavender
##"#dcbeff",
##Brown
#"#9A6324",
###Beige
##"#fffac8",
##Maroon
#"#800000",
##Mint
#"#aaffc3",
##Olive
#"#808000",
##Apricot
#"#ffd8b1",
##Navy
#"#000075",
##Grey
#"#a9a9a9",
###White
##"#ffffff",
##Black
#"#000000")

##Make a list of our species names
#species_names <- species_ids %>%
#        filter(name %in% tree$tip.label) %>%
#        pull(bacsort_species) %>%
#        unique() %>%
#        sort()
#
##Reduce the number of species colours
#species_cols <- species_cols[1:length(species_names)]
#
##Bind them to the colours
#names(species_cols) <- species_names

species_cols <- c("Citrobacter amalonaticus" =  "#e6194B",
"Citrobacter arsenatis" =     "#000000",
"Citrobacter braakii" =       "#4363d8",
"Citrobacter cronae" =        "#000000",
"Citrobacter europaeus" =     "#000000",
"Citrobacter farmeri" =       "#000000",
"Citrobacter freundii" =      "#f032e6",
"Citrobacter gillenii" =      "#bfef45",
"Citrobacter koseri" =        "#000000",
"Citrobacter murliniae" =     "#000000",
"Citrobacter pasteurii" =     "#9A6324",
"Citrobacter portucalensis" = "#800000",
"Citrobacter rodentium" =     "#000000",
"Citrobacter sedlakii" =      "#000000",
"Citrobacter telavivensis" =  "#000000",
"Citrobacter tructae" =       "#000000",
"Citrobacter unknown" =       "#a9a9a9",
"Citrobacter werkmanii" =     "#000000",
"Citrobacter youngae" =       "#dcbeff",
"Citrobacter meridianamericanus" = "#000000")

#Define our colours list for our collection names
collection_cols <- c("lightblue", "black", "darkred")
        
#Make a list of our species names
collection_names <- species_ids %>%
        filter(name %in% tree$tip.label) %>%
        pull(collection) %>%
        unique() %>%
        sort()

#Bind them to the colours
names(collection_cols) <- collection_names

#Filter our metadata
metadata <- metadata %>% filter(name %in% tree$tip.label)

#Remove unwanted species colours from the legend
species_cols <- species_cols[names(species_cols) %in% metadata$bacsort_species]

#Plot our tree
p <- ggtree(tree) %<+%
        metadata +
        geom_tiplab(size = 3,
                align = TRUE,
                linesize = 0.15,
                aes(color = bacsort_species)
                ) + theme(legend.position = "none")+
        geom_tippoint(size = 1.5,
                aes(colour = collection)
                ) + scale_fill_manual(
        aesthetics = c("colour", "fill"),
        values = c(species_cols, collection_cols),
        na.value = 'grey'
        )

species_coord <- fortify(p$data) %>%
        filter(isTip == TRUE) %>%
        select(label, bacsort_species, y) %>%
        arrange(y) %>%
        group_by(bacsort_species) %>%
        summarise(
                firstocc = first(label),
                lastocc = last(label),
                count = n()
        )

#Bind our colours and names
#Species_cols_df <- suppressWarnings(cbind(species_cols, species_names)) %>% as.data.frame()

#Rename our coloumns
#Species_cols_df <- Species_cols_df %>% rename(color = species_cols, bacsort_species = species_names)

#Convert our colours to a dataframe
Species_cols_df <- species_cols %>% as.data.frame() %>% rownames_to_column("bacsort_species") %>% rename("color" = ".")

#Join our df of coordinates for our annotations with our colours for our annotations
species_coord <- left_join(species_coord, Species_cols_df, by = "bacsort_species") %>% arrange(desc(count))

strip_fontsize <- 4
strip_offset <- 0.05
strip_text_offset <- 0.001
strip_barsize <- 2
strip_text_hjust <- 0

#Plot our core tree
p2 <- p + geom_strip(
                species_coord$firstocc[1],
                species_coord$lastocc[1],
                barsize = strip_barsize,
                color = species_coord$color[1],
                label = species_coord$bacsort_species[1],
                fontsize = strip_fontsize,
                offset = strip_offset,
                offset.text = strip_text_offset,
                hjust = strip_text_hjust) +
        geom_strip(
                species_coord$firstocc[2],
                species_coord$lastocc[2],
                barsize = strip_barsize,
                color = species_coord$color[2],
                label = species_coord$bacsort_species[2],
                fontsize = strip_fontsize,
                offset = strip_offset,
                offset.text = strip_text_offset,
                hjust = strip_text_hjust
        ) +
        geom_strip(
                species_coord$firstocc[3],
                species_coord$lastocc[3],
                barsize = strip_barsize,
                color = species_coord$color[3],
                label = species_coord$bacsort_species[3],
                fontsize = strip_fontsize,
                offset = strip_offset,
                offset.text = strip_text_offset,
                hjust = strip_text_hjust
        ) +
        geom_strip(
                species_coord$firstocc[4], 
                species_coord$lastocc[4],
                barsize = strip_barsize,
                color = species_coord$color[4],
                label = species_coord$bacsort_species[4],
                fontsize = strip_fontsize,
                offset = strip_offset,
                offset.text = strip_text_offset,
                hjust = strip_text_hjust
        ) +
        geom_strip(
                species_coord$firstocc[5],
                species_coord$lastocc[5],
                barsize = strip_barsize,
                color = species_coord$color[5],
                label = species_coord$bacsort_species[5],
                fontsize = strip_fontsize,
                offset = strip_offset,
                offset.text = strip_text_offset,
                hjust = strip_text_hjust
        ) +
        geom_strip(
                species_coord$firstocc[6],
                species_coord$lastocc[6],
                barsize = strip_barsize,
                color = species_coord$color[6],
                label = species_coord$bacsort_species[6],
                fontsize = strip_fontsize,
                offset = strip_offset,
                offset.text = strip_text_offset,
                hjust = strip_text_hjust
        ) +
        geom_strip(
                species_coord$firstocc[7],
                species_coord$lastocc[7],
                barsize = strip_barsize,
                color = species_coord$color[7],
                label = species_coord$bacsort_species[7],
                fontsize = strip_fontsize,
                offset = strip_offset,
                offset.text = strip_text_offset,
                hjust = strip_text_hjust
        )+ theme(legend.position = "none"
                )

p2 <- p2  + scale_x_continuous(limits = c(NA, 0.1825))
```

```{r, echo=FALSE, fig.height=8, fig.width=12}
#ggsave("images/Figure_1_Internal_tree_w_refs.pdf", p2, device = "pdf",  height = 8.27, width = 11.69)
ggsave("images/Figure_1_Internal_tree_w_refseq_refs.pdf", p2, device = "pdf",  height = 11.69, width = 8.27)
```
