---
title: "SKA analysis"
author: "Max Cummins"
date: "`r Sys.Date()`"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)

library(tidyverse, quietly = TRUE)
library(ggtree)
library(ape)
library(phangorn)
library(phytools)
```

# Outline

Here I will document the processing of SKA-borne SNP data for the investigation of relatedness between Citrobacter samples.

First lets read in our datasets:

```{r cars}

#Read in SNP distances
dists <- read_delim("results/ska_fastq_distance/Citrobacter_April.distances.tsv", 
    delim = "\t", escape_double = FALSE, 
    trim_ws = TRUE, show_col_types = FALSE)

#Read in SNP Clusters
clusts <- read_delim("results/ska_fastq_distance/Citrobacter_April.clusters.tsv", 
    delim = "\t", escape_double = FALSE, 
    trim_ws = TRUE, show_col_types = FALSE)

#Read in MLST data
mlst <- read_delim(
    "results/summaries/mlst.txt",
    delim = "\t",
    escape_double = FALSE,
    trim_ws = TRUE,
    show_col_types = FALSE
)

#Read in the list of genomes which passed QC
QC_pass <- read_csv("2022/delims/QC_pass.txt", col_names = "name", show_col_types = FALSE)

#Read in our metadata
metadata <- read_excel("Metadata_public_genomes.xlsx", 
    sheet = "metadata_only")

#Read in our species IDs
species_ids <- read_csv("2022/bacsort/Species_ids.csv")
```

## Editing of MLSTs

Next we need to process our MLST data so that novel STs are not treated as the
same but rather as different sequence types depending on their allelic profiles.

We will make some placeholder names in the mean time and change them later pre-
publication.

```{r }
#Filter out samples which didnt pass QC
mlst <- mlst %>% filter(name %in% QC_pass$name)

#Produce placeholder STs
placeholder_STs <- mlst %>% select(-name,-scheme) %>% unique() %>% filter(ST == "-") %>% select(-ST) %>%
       # filter(
       #         !grepl("-", allele1),
       #         !grepl("-", allele2),
       #         !grepl("-", allele3),
       #         !grepl("-", allele4),
       #         !grepl("-", allele5),
       #         !grepl("-", allele6),
       #         !grepl("-", allele7)) %>% 
mutate(ST_new = paste0("placeholder_ST",as.character(1:nrow(.)))) %>% mutate(allele_cat = paste0(allele1, allele2, allele3, allele4, allele5, allele6, allele7)) %>% select(ST_new, allele_cat)

#Produce a list of the mlsts within the collection
ST_list <- mlst %>% mutate(allele_cat = paste0(allele1, allele2, allele3, allele4, allele5, allele6, allele7)) %>% left_join(placeholder_STs, by = "allele_cat") %>% select(-allele_cat) %>% mutate(ST_new = if_else(is.na(ST_new), ST, ST_new))
```


Now we will assign sources based on our sample names

```{r }
#Create a new column for our collection_names
species_ids <- species_ids %>%
        mutate(collection = case_when(grepl("GCF", name) ~ "RefSeq",
                                      grepl("SD", name) ~ "Gull",
                                      grepl("SG", name) ~ "Gull",
                                      grepl("ERQ", name) ~ "Wastewater",
                                      grepl("DB", name) ~ "Wastewater"))
                                      


#Create our metadata sheet (we will expand on this later with extra data)
metadata <- left_join(species_ids, metadata)

#Fix the source column for internal isolates
metadata <- metadata %>% mutate(Source = if_else(collection == "RefSeq", Source, collection)) %>% rename("ID" = name)



#Combine our metadata and our clusters for use in microreact
data_microreact <- left_join(clusts, metadata, by.x = "ID", by.y = "name")

#Combine our ST data and metadata/clusters for use in microreact
data_microreact <- ST_list %>% select(name, ST_new) %>% rename("ID" = name) %>% right_join(data_microreact)

#Write this to file
write_delim(x = data_microreact, file = "results/ska_fastq_distance/Citrobacter_April.clusters_w_metadata.tsv", delim = "\t")



data_microreact1 <- data_microreact %>% rename("Sample 1" = "ID") %>% rev()

data_microreact2 <- data_microreact %>% rename("Sample 2" = "ID")

dists2 <- inner_join(data_microreact1, dists, by = "Sample 1")

dists2 <- inner_join(dists2, data_microreact2, by = "Sample 2")


dists3 <- dists2 %>% filter(SNPs < 1000) %>% drop_na(ST_new.x, ST_new.y) %>% filter(grepl("SD|ERQ|SG", `Sample 1`) | grepl("SD|ERQ|SG", `Sample 2`))


```


```{r}
#Set the path for the gull overlap tree
ww_overlap_tree_path <- "results/summaries/trees/mashtree_citro_ww_overlap.tree"

#Read in our tree
ww_tree <- read.tree(ww_overlap_tree_path)

#Midpoint root our tree
ww_tree <- midpoint.root(ww_tree)

dists2 %>% filter(`Sample 1` %in% ww_tree$tip.label) %>% filter(`Sample 2` %in% ww_tree$tip.label) %>% View()
```


```{r}

#Set the path for the gull overlap tree
gull_overlap_tree_path <- "results/summaries/trees/mashtree_citro_gull_overlap.tree"

#Read in our tree
gull_tree <- read.tree(gull_overlap_tree_path)

#Midpoint root our tree
gull_tree <- midpoint.root(gull_tree)

#Check for gull matches
dists2 %>% filter(`Sample 1` %in% gull_tree$tip.label) %>% filter(`Sample 2` %in% gull_tree$tip.label) %>% View()

```

